org.eclipse.scanning.api/src/org/eclipse/scanning/api/device/DeviceRequestHandler.java

```diff
@@ -1,249 +1,249 @@
+ *******************************************************************************/
+package org.eclipse.scanning.api.device;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+
+import java.util.Collection;
+
+import org.eclipse.scanning.api.INameable;
+import org.eclipse.scanning.api.IScannable;
+import org.eclipse.scanning.api.ITerminatable;
+import org.eclipse.scanning.api.ModelValidationException;
+import org.eclipse.scanning.api.MonitorRole;
+import org.eclipse.scanning.api.annotation.ui.DeviceType;
+import org.eclipse.scanning.api.event.EventException;
+import org.eclipse.scanning.api.event.core.IPublisher;
+import org.eclipse.scanning.api.event.core.IRequestHandler;
+import org.eclipse.scanning.api.event.scan.DeviceAction;
+import org.eclipse.scanning.api.event.scan.DeviceInformation;
+import org.eclipse.scanning.api.event.scan.DeviceRequest;
+import org.eclipse.scanning.api.malcolm.attributes.IDeviceAttribute;
+import org.eclipse.scanning.api.scan.ScanningException;
+
+/**
+ * TODO FIXME Is the idea of having request/response calls correct for exposing
+ * services on the server to the client?
+ * Disadvantages:
+ * 1. Pushes to much to client because it has make/receive request/responses.
+ * 2. Lot of specific code around a given service in the client. Would be nicer to discover services
+ *    using the original service directly, but it going back to the server...
+ *
+ * Advantages:
+ * 1. Can work for any client, python and javascript included
+ * 2. No JSON serialization issues
+ *
+ *
+ * @author Matthew Gerring
+ *
+ */
+public class DeviceRequestHandler implements IRequestHandler<DeviceRequest> {
+
+	private IRunnableDeviceService    dservice;
+	private DeviceRequest             bean;
+	private IPublisher<DeviceRequest> publisher;
+	private IScannableDeviceService   cservice;
+
+	public DeviceRequestHandler(IRunnableDeviceService  dservice,
+			              IScannableDeviceService cservice,
+			              DeviceRequest           bean,
+			              IPublisher<DeviceRequest> statusNotifier) {
+
+		this.dservice = dservice;
+		this.cservice = cservice;
+		this.bean     = bean;
+		this.publisher = statusNotifier;
+	}
+
+	@Override
+	public DeviceRequest getBean() {
+		return bean;
+	}
+
+	@Override
+	public IPublisher<DeviceRequest> getPublisher() {
+		return publisher;
+	}
+
+	@Override
+	public DeviceRequest process(DeviceRequest request) {
+		try {
+			if (request.getDeviceType()==DeviceType.SCANNABLE) {
+				processScannables(request, cservice);
+			} else {
+				processRunnables(request, dservice);
+			}
+			return request;
+
+		} catch (ModelValidationException ne) {
+			DeviceRequest error = new DeviceRequest();
+			error.merge(request);
+			error.setErrorMessage(ne.getMessage());
+			error.setErrorFieldNames(ne.getFieldNames());
+			return error;
+
+		} catch (Exception ne) {
+			ne.printStackTrace();
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use a logger to log this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1148)